packages/react-instantsearch-core/src/connectors/connectInfiniteHits.js

```diff
@@ -45,39 +47,63 @@ export default createConnector({
     const results = getResults(searchResults, this.context);
 
     this._allResults = this._allResults || [];
-    this._previousPage = this._previousPage || 0;
+    this._prevState = this._prevState || {};
 
     if (!results) {
       return {
         hits: [],
+        hasPrevious: false,
         hasMore: false,
+        refine: () => {},
+        refinePrevious: () => {},
+        refineNext: () => {},
       };
     }
 
-    const { hits, hitsPerPage, page, nbPages } = results;
+    const {
+      page,
+      hits,
+      hitsPerPage,
+      nbPages,
+      _state: { page: p, ...currentState } = {},
+    } = results;
 
     const hitsWithPositions = addAbsolutePositions(hits, hitsPerPage, page);
     const hitsWithPositionsAndQueryID = addQueryID(
       hitsWithPositions,
       results.queryID
     );
 
-    if (page === 0) {
-      this._allResults = hitsWithPositionsAndQueryID;
-    } else if (page > this._previousPage) {
+    if (
+      this._firstReceivedPage === undefined ||
+      !isEqual(currentState, this._prevState)
+    ) {
+      this._allResults = [...hitsWithPositionsAndQueryID];
+      this._firstReceivedPage = page;
+      this._lastReceivedPage = page;
+    } else if (this._lastReceivedPage < page) {
       this._allResults = [...this._allResults, ...hitsWithPositionsAndQueryID];
-    } else if (page < this._previousPage) {
-      this._allResults = hitsWithPositionsAndQueryID;
+      this._lastReceivedPage = page;
+    } else if (this._firstReceivedPage > page) {
+      this._allResults = [...hitsWithPositionsAndQueryID, ...this._allResults];
+      this._firstReceivedPage = page;
     }
 
+    this._prevState = currentState;
+
+    const hasPrevious = this._firstReceivedPage > 0;
     const lastPageIndex = nbPages - 1;
     const hasMore = page < lastPageIndex;
-
-    this._previousPage = page;
+    const refinePrevious = () => this.refine(this._firstReceivedPage - 1);
+    const refineNext = () => this.refine(this._lastReceivedPage + 1);
 
     return {
       hits: this._allResults,
+      hasPrevious,
       hasMore,
+      refine: refineNext,
```

Comment:
The `refine` function is overridden by [`createConnector`](https://github.com/algolia/react-instantsearch/blob/10ae0a0c31664cf941de9b6d9688cc6903678ca3/packages/react-instantsearch-core/src/core/createConnector.js#L336-L341). The outcome is that `refine` is different than `refineNext` for the connected component. `refine` uses the `currentRefinement + 1` but the `refineNext` function uses the `_lastReceivedPage + 1`. It depends on some condition but the call to `refine` is a noop if we previously click on `refinePrevious`.

Reply:
Oh, I though I managed this case but I missed some details. I pushed a fix here https://github.com/algolia/react-instantsearch/pull/2296/commits/c91b8c0dbd7af0289148d32612f945c0df008442#diff-c3b7a7fe9c467c6bbe02a87f5b8729c4. Let me know if it's ok for you.

Reply:
The issue persist the two function does not have the same behaviour.

![issue](https://user-images.githubusercontent.com/6513513/56652614-12e59080-668c-11e9-8808-51c7a6db6115.gif)

Reply:
Rah, should be ok now. Thanks for pointing that out.

I also added a "with custom rendering" story to ease testing.

Reply:
Rah, should be ok now. Thanks for pointing that out.

I also added a "with custom rendering" story to ease testing.